jans-cedarling/cedarling/src/authz/entities/mod.rs

```diff
@@ -97,17 +107,104 @@ pub fn create_resource_entity(
     })?;
 
     let parsed_typename = parse_namespace_and_typename(&resource.resource_type);
-    // fetch the schema record from the json-schema.
-    let schema_record = schema
-        .entity_schema_record(&parsed_typename.namespace(), parsed_typename.typename)
-        .ok_or(CedarPolicyCreateTypeError::CouldNotFindEntity(
-            entity_uid.to_string(),
-        ))?;
 
     Ok(create_entity(
         entity_uid,
-        &parsed_typename.namespace(),
-        schema_record,
+        &parsed_typename,
+        schema,
         &resource.payload.into(),
+        HashSet::new(),
     )?)
 }
+
+/// Describe errors on creating role entity
+#[derive(thiserror::Error, Debug)]
+pub enum RoleEntityError {
+    #[error("could not create Jans::Role entity from {token_kind} token: {error}")]
+    Create {
+        error: CedarPolicyCreateTypeError,
+        token_kind: TokenKind,
+    },
+}
+
+/// Create `Role` entity from based on `TrustedIssuer` or default value of `RoleMapping`
+pub fn create_role_entities(
+    schema: &CedarSchemaJson,
+    tokens: &DecodeTokensResult,
+) -> Result<Vec<cedar_policy::Entity>, RoleEntityError> {
+    // get role mapping or default value
+    let role_mapping = tokens
+        .trusted_issuer
+        .map(|trusted_issuer| trusted_issuer.get_role_mapping().unwrap_or_default())
+        .unwrap_or_default();
+
+    let role_entity_type: &str = "Jans::Role";
+    let parsed_typename = parse_namespace_and_typename(role_entity_type);
+
+    // map payload from token
+    let token_data: &'_ TokenPayload = match role_mapping.kind {
+        TokenKind::Access => &tokens.access_token,
+        TokenKind::Id => &tokens.id_token,
+        TokenKind::Userinfo => &tokens.userinfo_token,
+    };
+
+    // get payload of role id in JWT token data
+    let Ok(payload) = token_data.get_payload(role_mapping.role_mapping_field) else {
+        // if key not found we return empty vector
+        return Ok(Vec::new());
+    };
+
+    // it can be 2 scenario when field is array or field is string
```

Comment:
Maybe it would be easier to have something like
```
#[serde(untagged)]
enum EntityValue { Single(EntityUid), Multiple(Vec<EntityUid>) }
```
and then just use `serde_json::from_value` to decode each value in `TokenPayload::payload`.

Reply:
TokenPayload it is just HashMap of json values in JWT. We should not hardcode the posible values types. Because we build `cedar-policy` entities based on the `cedar-policy::Schema`. So we do it dynamically. Or I didn't get.

Reply:
This commit on a spare branch explains it better: https://github.com/JanssenProject/jans/commit/c8c12ecaa2418765b5f999aed93ca176b2e33e36

Use it if you like.